Add State management MVI Post (#203)
godaddy/engineering: _posts/2021-11-08-android-state-management-mvi.md

---
layout: post
title: GoDaddy Studio’s Journey with State Management and MVI / Unidirectional Data Flow on Android
date: 2021-11-05 00:00:00 -0700
cover: /assets/images/android-state-management-mvi/cover.jpg
options:
  - full-bleed-cover
excerpt: In this post, we will look at the journey that the GoDaddy Studio Android team took with how UI state is managed across the app. We will cover MVVM and how it caused issues, the initial MVI implementation and the issues we faced. Finally, we will look at how we landed on using Spotify’s Mobius Framework for managing state. 
keywords: Android, Architecture, MVI, MVVM, State machine, Mobius
authors:
  - name: Rebecca Franks
    title: Software Development Engineer IV
    url: https://www.twitter.com/riggaroo
    photo: /assets/images/rebecca_franks.jpg
  - name: Kamil Slesinski
    title: Software Development Engineer IV
    url: https://www.linkedin.com/in/kamilslesinski/
    photo: /assets/images/kamil_slesinski.jpeg
---

## History ⌛️

> On the GoDaddy Studio Android app, we started out with a fresh codebase around the end of 2018. Right when Kotlin was starting to become the popular choice and the community had moved on from choosing MVP as the preferred architecture choice, to MVVM. We decided to adopt MVVM where it made sense, but we were intrigued by the prospects of MVI/[Unidirectional data flow](https://en.wikipedia.org/wiki/Unidirectional_Data_Flow_(computer_science)). After watching a few presentations on the topic, we proceeded to design our Canvas Editor screen using a homegrown version of MVI, very similar to the [code](https://github.com/oldergod/android-architecture) described in this [presentation](https://www.youtube.com/watch?v=64rQ9GKphTg&ab_channel=oldergod).

Reviewer: ```suggestion
The GoDaddy Studio Android app began with a fresh codebase around the end of 2018. This was around the same time Kotlin was starting to become a popular choice and the community had moved on from choosing MVP as the preferred architecture choice, to MVVM. We decided to adopt MVVM where it made sense, but we were intrigued by the prospects of MVI/[Unidirectional data flow](https://en.wikipedia.org/wiki/Unidirectional_Data_Flow_(computer_science)). After watching a few presentations on the topic, we proceeded to design our Canvas Editor screen using a homegrown version of MVI, very similar to the [code](https://github.com/oldergod/android-architecture) described in this [presentation](https://www.youtube.com/watch?v=64rQ9GKphTg&ab_channel=oldergod).
```

As time went on, we started facing issues with the MVI approach we used. The way we implemented it, unfortunately allowed for race conditions and caused a lot of headaches for the team. It was at this point that we decided to investigate using existing frameworks for unidirectional data flow, and see if they could solve the problems we were seeing with ours. After evaluating a few, we settled on using the [Spotify Mobius Framework](https://github.com/spotify/mobius) for various reasons which we will dive into in this blog post.

## State Management 

> Let’s step back a bit and learn about the issues we faced with our approaches over the past couple of years. What does it mean when we talk about State Management for UI? State management refers to keeping track of how a user interface should look and how it should react to different user inputs. On a complex screen with plenty of buttons, gestures and text input - managing the state is a full time concern. 

Author: ```suggestion
Let’s step back a bit and learn about the issues we faced with our approaches over the past couple of years. What does it mean when we talk about State Management for UI? State management refers to keeping track of how a user interface should look and how it should react to different user inputs. On a complex screen with plenty of buttons, gestures and text input - managing the state is a full-time concern. 
```

Take this screen for example: 
{:refdef: style="text-align: center;"}
![]({{site.baseurl}}/assets/images/android-state-management-mvi/OverallState.gif){: width="250" }
{: refdef}
There are plenty of concerns that need to be managed and tracked on this screen alone:

- What should happen when the project is loading up?
- Has an error occurred?
- What is the currently selected tool on screen?
- What are the layers and properties that should be displayed in this project?

Not only is there the current state of what should be rendered on screen, but there are many different interactions that need to mutate this state - scaling gestures, button taps, colour changes etc. 

## MVVM and the issues we faced 😣

> A couple of years ago, the Android community began advocating for using **some kind of architecture,** as we all realized that placing all your logic inside the Activity, was a perfect recipe for the messiest spaghetti bolognaise you could imagine (delicious but not very elegant).

Author: Apparently they are both right 😬  https://www.newgrounds.com/bbs/topic/1043082 haha

> First came **MVP** (Model-View-Presenter) and there were great benefits to applying this architecture to our apps: we got testability, separation of concerns and the ability to re-use logic on other screens. 

Author: This looks like both are correct - https://www.grammarly.com/blog/comma-before-and/ just need to be consistent. Will stick without the ,

![I’m sure we’ve all seen or made a screen like this before 😬]({{site.baseurl}}/assets/images/android-state-management-mvi/mvvm_issues_overlapping.png){: width="250" }
{: refdef}

There are several drawbacks when using this approach of MVVM with separate LiveData objects:

- It is hard to know at a single point in time, what the UI should look like, if there are multiple LiveData objects that change the state of the screen.
- There is no single snapshot to be able to recreate the UI from easily.
- This also then begs the question, are we handling all the potential cases here?

One way to potentially improve on this behaviour, is to use data classes that contain the state information in one class, and expose only a singular LiveData object to the UI. This would help solve the issue of having multiple observables emitting different state that we’d need to keep track of, and its a building block of how MVI can work, as we will explore next. 

# MVI / Unidirectional Data Flow

Now that we understand where potential issues can come from with MVVM and multiple LiveData observables, let’s talk a bit about Unidirectional Data Flow or MVI (Model-View-Intent) and our experience with it.

> With Unidirectional Data Flow everything is modelled around a single state object. This state is immutable, to prevent unwanted side effects. There is a clear flow of how data changes through the system, since it only happens in a single direction.

Author: Another US vs British english thing apparently - https://www.grammarly.com/blog/modeling-or-modelling/

{:refdef: style="text-align: center;"}
<figure>

>   <img src="{{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>

Reviewer 2: ```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>
```
Jekyll was throwing a Liquid syntax error on this. Also, any chance this alt text can be updated?

Author: ```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="Diagram showing Unidirectional Data Flow"/>
```

- It is an open-sourced, maintained, production-ready framework with support for Android and RxJava
- The framework forces you to design business logic with consistent state and well-defined events (we have more than 200 unique events on the editor screen!)
- It makes it much easier to reason about app behaviour when you have this kind of structure in place
- We now have a standardized ViewModel implementation across the app

With every new framework or architecture decision, there is likely never going to be the solution that fits everyone, and adopting something like the Mobius Framework also comes with these disadvantages to adopting:

- It requires quite a bit of boilerplate code to set up. We’ve solved this using shared code templates to generate most of the classes we need. 
- New framework for people to learn. Any new engineer needs to spend a bit of time learning about the framework, working through some examples and implementing a feature with it. 

# Summary 🎉

> We’ve learnt a lot over the past couple of years, and I don’t doubt that we won’t learn more in the future. 

Author: https://www.grammarly.com/blog/learned-learnt/ both are right again 😄  Leaving as learnt.
